File: Projetos_C/Terceiro Semestre/Estrutura de Dados/Aula 11/Ex6.c
#include<stdio.h>
#include<stdlib.h>

#define QUEUESIZE 10

struct queue {
	int front, rear;
	char items[QUEUESIZE];
};

int empty (struct queue *pq);
char removeq (struct queue *pq);
void insert (struct queue *pq, char x);
void print (struct queue *ps);
void posicaoZero(struct queue *pq);

 

int main (){
    struct queue caracteres[QUEUESIZE];
    caracteres->front=0;
    caracteres->rear=-1;
    for(int i=0; i<QUEUESIZE; i++)
        insert(caracteres, 'a'+i);

    for(int i=0; i<5; i++)
        removeq(caracteres);
    
    insert(caracteres, 'z');
    print(caracteres);
	
	return 0;
}

void print (struct queue *ps){
	for (int i = ps->front; i<=ps->rear; i++)
		printf("%c\n", ps->items[i]);
}


int empty (struct queue *pq){
	if (pq->rear < pq->front)
		return 1;
	return 0;	
}

char removeq (struct queue *pq){
	char x;

	if (empty(pq)){
		printf("ERRO- FILA VAZIA");
		exit(1);
	}

	x=pq->items[pq->front];
	pq->front++;

	return x;
}

void insert(struct queue *pq, char x) {
    if (pq->rear == QUEUESIZE - 1) {
        if (pq->front == 0) {
            printf("ERRO- FILA CHEIA");
            exit(1);
        } else {
            posicaoZero(pq);
        }
    }
    pq->rear++;
    pq->items[pq->rear] = x;
}

void posicaoZero(struct queue *pq) {
    for (int i = pq->front; i <= pq->rear; i++) {
        pq->items[i - pq->front] = pq->items[i];
    }
    pq->rear = pq->rear - pq->front;
    pq->front = 0;
}

File: Projetos_Python/3 Semestre/Primeira Fase/Prova1/3-Pilha e Fila/Lista1/Ex6.py
"""
6. Se um vetor armazenando uma fila não é considerado circular, pode ocasionar o problema de
overflow quando a variável rear chega ao fim da fila, no entanto, existem espaços anteriores à 
variável front, que poderia armazenar novos elementos.

Uma possível solução é deslocar para baixo todos elementos da uma fila a cada remoção. Porém 
esta solução é custosa. Um método menos custoso é adiar o deslocamento até que rear seja igual ao 
último índice do vetor. Quando essa situação ocorre e faz-se a tentativa de inserir um elemento na 
fila, a fila inteira é deslocada para baixo, de modo que o primeiro elemento da fila fique na posição 
0 do vetor. Escreva as rotinas remove, insert e empty usando esse método
"""

QUEUESIZE=7
class Fila:
    def __init__(self):
        self.front=0
        self.rear=-1
        self.valores = [None]*QUEUESIZE

    def fila_vazia(self):
        if (self.rear<self.front):
            return True
        return False
    
    def fila_cheia(self):
        if(self.rear==QUEUESIZE-1):
            return True
        return False

    
    def remove(self):
        if(self.fila_vazia()):
            print("Fila Vazia")
            exit(1)
        primeiro=self.valores[self.front]
        self.front+=1
        return primeiro
    
    def insert (self, valor):
        if(self.fila_cheia()):
            if self.front==0:
                print("Erro - Fila Cheia")
                exit(1)
            else:
                self.posicaoZero()
                
        self.rear+=1
        self.valores[self.rear]=valor

    def posicaoZero(self):
        for i in range(self.front, self.rear+1):
            self.valores[self.front-i]=self.valores[i]
            self.valores[i]=0
        self.rear=self.rear-self.front
        self.front=0

teste=Fila()

for i in range(0,7):
    teste.insert(i)
print(teste.valores)
for i in range(0,4):
    teste.remove()
print(teste.valores)
teste.insert('8')

print(teste.valores)

    



    
    